docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,137 @@
+
+<details><summary>lType library methods (dropdown)</summary>
+Note: methods under a different type heading (ex: `Singleton`) imply that the methods are only available for those types. At the implementation level, there is a check to make sure that the type-specific methods are being called on the correct types (e.g, for `getIndexer()`, assert that `isTable()` is true).
+
+#### Any
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `isString()` | `boolean` | returns true if self is of type `string` |
+| `isNumber()` | `boolean` | returns true if self is of type `number` |
+| `isBoolean()` | `boolean` | returns true if self is of type `boolean` |
+| `isTable()` | `boolean` | returns true if self is of type `table` |
+| `isThread()` | `boolean` | returns true if self is of type `thread` |
+| `isFunction()` | `boolean` | returns true if self is of type `function` |
+| `isBuffer()` | `boolean` | returns true if self is of type `buffer` |
+| `isNil()` | `boolean` | returns true if self is of type `nil` |
+| `isClass()` | `boolean` | returns true if self is of type `class` (do we need this?) |
+| `isBooleanSingleton()` | `boolean` | returns true if self is a boolean singleton |
+| `isStringSingleton()` | `boolean` | returns true if self is a string singleton |
+| `isA(arg: lType)` | `boolean` | returns true if arg is the same type as self |
+
+#### Primitive
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getType()` | `string` | returns either "nil", "boolean", "string", "thread", "function", "table", or "buffer" |
+
+#### Singleton
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getValue()` | `string` | returns either "true", "false", or a string singleton |
+
+#### Table
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getProps()` | `table` | returns a type representation of tables (e.g. {name = "John"} will return {[string] = "string"}) |
+| `getIndexer()` | `table` | returns a type representation of arrays (e.g. {1, "hi", 3} will return {[number] = "number" \| "string"}) |
+
+#### Class
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getName()` | `table` | returns the name of self's class |
+| `getParent()` | `lType` | returns lType userdata of self's parent |
+
+</details>
+
+### Implementation
```

Comment:
Can probably be left out of the RFC, in my opinion.

Reply:
There's a better implementation section that could be written that actually describes the interesting work here (implementing a `user` type function into the builtins that calls serialization and deserialization code and interacts with the VM itself to run user code in an established "type function" environment with the right libraries available and all of the type aliases and functions from the code available).

Reply:
In a recent commit, I've revised the implementation section to leave out specific details and include more details on the interactions with the VM